File: CygniParser/src/cygni/lexer/Lexer.java
package cygni.lexer;

import cygni.exceptions.LexicalException;

import java.util.ArrayList;
import java.util.HashMap;

public class Lexer {
    public ArrayList<Token> tokens;
    private String code;
    private int line;
    private int col;
    private int offset;
    private StringBuilder builder;
    private static HashMap<String, Tag> operators;

    static {
        operators = new HashMap<String, Tag>();
        operators.put("+", Tag.Add);
        operators.put("-", Tag.Subtract);
        operators.put("*", Tag.Multiply);
        operators.put("/", Tag.Divide);
        operators.put("%", Tag.Modulo);
        operators.put(">", Tag.GreaterThan);
        operators.put("<", Tag.LessThan);
        operators.put(">=", Tag.GreaterThanOrEqual);
        operators.put("<=", Tag.LessThanOrEqual);
        operators.put("==", Tag.Equal);
        operators.put("!=", Tag.NotEqual);
        operators.put("(", Tag.LeftParenthesis);
        operators.put(")", Tag.RightParenthesis);
        operators.put("[", Tag.LeftBracket);
        operators.put("]", Tag.RightBracket);
        operators.put("{", Tag.LeftBrace);
        operators.put("}", Tag.RightBrace);
        operators.put(":", Tag.Colon);
        operators.put(",", Tag.Comma);
        operators.put(".", Tag.Dot);
        operators.put(";", Tag.Semicolon);
        operators.put("=", Tag.Assign);
        operators.put("->", Tag.RightArrow);
    }

    public Lexer(String code) {
        this.code = code;
        this.line = 0;
        this.col = 0;
        this.offset = 0;
        this.builder = new StringBuilder();
    }

    public ArrayList<Token> readAll() throws LexicalException {
        ArrayList<Token> tokens = new ArrayList<Token>();
        skipWhitespaces();
        while (!eof()) {
            if (Character.isDigit(peek())) {
                tokens.add(readInt());
            } else if (peek() == '"' || peek() == '\'') {
                tokens.add(readString());
            } else if (Character.isLetter(peek()) || peek() == '_') {
                tokens.add(readIdentifier());
            } else if ("+-*/%><=!()[]{}:,.;".indexOf(peek()) >= 0) {
                tokens.add(readOperator());
            } else {
                throw new LexicalException(line, col, "unsupported token");
            }
            skipWhitespaces();
        }
        tokens.add(new Token(line, col, Tag.Eof, "<EOF>"));
        return tokens;
    }

    private Token readInt() throws LexicalException {
        reset();
        while (!eof() && Character.isDigit(peek())) {
            consume();
        }
        if (eof()) {
            return new Token(line, col, Tag.Integer, Integer.parseInt(builder.toString()));
        } else {
            if (peek() == '.') {
                consume();
                return readFloat();
            } else {
                return new Token(line, col, Tag.Integer, Integer.parseInt(builder.toString()));
            }
        }
    }

    private Token readFloat() throws LexicalException {
        while (!eof() && Character.isDigit(peek())) {
            consume();
        }
        if (eof()) {
            return new Token(line, col, Tag.Float, Double.parseDouble(builder.toString()));
        } else {
            if (peek() == 'E' || peek() == 'e') {
                consume();
                if (peek() == '+' || peek() == '-') {
                    consume();
                }
                if (eof() || !Character.isDigit(peek())) {
                    throw new LexicalException(line, col, "float literal");
                } else {
                    while (!eof() && Character.isDigit(peek())) {
                        consume();
                    }
                    return new Token(line, col, Tag.Float, Double.parseDouble(builder.toString()));
                }
            } else {
                return new Token(line, col, Tag.Float, Double.parseDouble(builder.toString()));
            }
        }
    }

    private Token readString() throws LexicalException {
        reset();
        forward();
        while (!eof() && peek() != '"') {
            if (peek() == '\\') {
                forward();
                if (eof()) {
                    throw new LexicalException(line, col, "string literal");
                } else {
                    builder.append(unescpaedChar(peek()));
                    forward();
                }
            } else {
                consume();
            }
        }
        if (eof()) {
            throw new LexicalException(line, col, "string literal");
        } else {
            forward();
            return new Token(line, col, Tag.String, builder.toString());
        }
    }

    private char unescpaedChar(char c) throws LexicalException {
        switch (c) {
            case 'b':
                return '\b';
            case 'n':
                return '\n';
            case 't':
                return '\t';
            case 'r':
                return '\r';
            case 'f':
                return '\f';
            case '\"':
                return '\"';
            case '\'':
                return '\'';
            case '\\':
                return '\\';
            default:
                throw new LexicalException(line, col, "unsupported escaped character");
        }
    }

    private Token readIdentifier() {
        reset();
        consume();
        while (!eof() && isIdentifierChar(peek())) {
            consume();
        }
        return new Token(line, col, Tag.Identifier, builder.toString());
    }

    private Token readOperator() throws LexicalException {
        char c1 = peek();
        String s1 = Character.toString(c1);
        forward();
        if (eof()) {
            if (operators.containsKey(s1)) {
                return new Token(line, col, operators.get(s1), s1);
            } else {
                throw new LexicalException(line, col, "operator literal");
            }
        } else {
            char c2 = peek();
            String s12 = Character.toString(c1) + Character.toString(c2);
            if (operators.containsKey(s12)) {
                forward();
                return new Token(line, col, operators.get(s12), s12);
            } else if (operators.containsKey(s1)) {
                return new Token(line, col, operators.get(s1), s1);
            } else {
                throw new LexicalException(line, col, "operator literal");
            }
        }
    }

    private void skipWhitespaces() {
        while (!eof() && Character.isWhitespace(peek())) {
            forward();
        }
    }

    private boolean isIdentifierChar(char c) {
        return Character.isLetterOrDigit(c) || c == '_';
    }

    private boolean eof() {
        return offset >= code.length();
    }

    private void forward() {
        if (code.charAt(offset) == '\n') {
            offset++;
            line++;
            col = 0;
        } else {
            offset++;
            col++;
        }
    }

    private char peek() {
        return code.charAt(offset);
    }

    private void consume() {
        builder.append(peek());
        forward();
    }

    private void reset() {
        builder.setLength(0);
    }
}

File: 2019-09-28/CygniCompiler/src/sources/Lexer.cpp
#include "Lexer.hpp"
#include <iostream>

Lexer::Lexer(String code) : code{std::move(code)}, line{1}, column{1}, offset{0}
{
}

std::vector<Token> Lexer::ReadAll()
{
    std::vector<Token> tokens;
    static String opChars = "+-*/%><=!()[]{}:,.;";
    SkipWhitespaces();
    while (!IsEof())
    {
        if (IsDigit(Peek()))
        {
            tokens.push_back(ReadInt());
        }
        else if (Peek() == '"' || Peek() == '\'')
        {
            tokens.push_back(ReadString());
        }
        else if (IsLetter(Peek()) || Peek() == '_')
        {
            tokens.push_back(ReadIdentifier());
        }
        else if (opChars.find(Peek()) != String::npos)
        {
            tokens.push_back(ReadOperator());
        }
        else
        {
            throw LexicalException(line, column, "unsupported token");
        }
        SkipWhitespaces();
    }
    tokens.emplace_back(line, column, Tag::Eof, "<EOF>");
    return tokens;
}

std::unordered_map<String, Tag> Lexer::operators = {
        {"+",  Tag::Add},
        {"-",  Tag::Subtract},
        {"*",  Tag::Multiply},
        {"/",  Tag::Divide},
        {"%",  Tag::Modulo},
        {">",  Tag::GreaterThan},
        {"<",  Tag::LessThan},
        {">=", Tag::GreaterThanOrEqual},
        {"<=", Tag::LessThanOrEqual},
        {"==", Tag::Equal},
        {"!=", Tag::NotEqual},
        {"(",  Tag::LeftParenthesis},
        {")",   Tag::RightParenthesis},
        {"[",  Tag::LeftBracket},
        {"]",  Tag::RightBracket},
        {"{",  Tag::LeftBrace},
        {"}",  Tag::RightBrace},
        {":",  Tag::Colon},
        {",",  Tag::Comma},
        {".",  Tag::Dot},
        {";",  Tag::Semicolon},
        {"=",  Tag::Assign},
        {"->", Tag::RightArrow},
        {"<:", Tag::UpperBound},
        {":>", Tag::LowerBound}
};

Token Lexer::ReadInt()
{
    Reset();
    while (!IsEof() && IsDigit(Peek()))
    {
        Consume();
    }
    if (IsEof())
    {
        return Token(line, column, Tag::Integer, builder);
    }
    else
    {
        if (Peek() == '.')
        {
            Consume();
            return ReadFloat();
        }
        else
        {
            return Token(line, column, Tag::Integer, builder);
        }
    }
}

Token Lexer::ReadFloat()
{
    while (!IsEof() && IsDigit(Peek()))
    {
        Consume();
    }
    if (IsEof())
    {
        return Token(line, column, Tag::Float, builder);
    }
    else
    {
        if (Peek() == 'E' || Peek() == 'e')
        {
            Consume();
            if (Peek() == '+' || Peek() == '-')
            {
                Consume();
            }
            if (IsEof() || !IsDigit(Peek()))
            {
                throw LexicalException(line, column, "float literal");
            }
            else
            {
                while (!IsEof() && IsDigit(Peek()))
                {
                    Consume();
                }
                return Token(line, column, Tag::Float, builder);
            }
        }
        else
        {
            return Token(line, column, Tag::Float, builder);
        }
    }
}

Token Lexer::ReadString()
{
    Reset();
    Forward();
    while (!IsEof() && Peek() != '"')
    {
        if (Peek() == '\\')
        {
            Forward();
            if (IsEof())
            {
                throw LexicalException(line, column, "string literal");
            }
            else
            {
                builder.push_back(UnescapedChar(Peek()));
                Forward();
            }
        }
        else
        {
            Consume();
        }
    }
    if (IsEof())
    {
        throw LexicalException(line, column, "string literal");
    }
    else
    {
        Forward();
        return Token(line, column, Tag::String, builder);
    }
}

uint16_t Lexer::UnescapedChar(uint16_t c)
{
    switch (c)
    {
        case 'b':
            return '\b';
        case 'n':
            return '\n';
        case 't':
            return '\t';
        case 'r':
            return '\r';
        case 'f':
            return '\f';
        case '\"':
            return '\"';
        case '\'':
            return '\'';
        case '\\':
            return '\\';
        default:
            throw LexicalException(line, column, "unsupported escaped character");
    }
}

Token Lexer::ReadIdentifier()
{
    Reset();
    Consume();
    while (!IsEof() && IsIdentifierChar(Peek()))
    {
        Consume();
    }
    return Token(line, column, Tag::Identifier, builder);
}

Token Lexer::ReadOperator()
{
    Char c1 = Peek();
    String s1;
    s1.push_back(c1);
    Forward();
    if (IsEof())
    {
        if (operators.find(s1) != operators.end())
        {
            return Token(line, column, operators[s1], s1);
        }
        else
        {
            throw LexicalException(line, column, "operator literal");
        }
    }
    else
    {
        Char c2 = Peek();
        String s12;
        s12.push_back(c1);
        s12.push_back(c2);
        if (operators.find(s12) != operators.end())
        {
            Forward();
            return Token(line, column, operators[s12], s12);
        }
        else if (operators.find(s1) != operators.end())
        {
            return Token(line, column, operators[s1], s1);
        }
        else
        {
            throw LexicalException(line, column, "operator literal");
        }
    }
}

void Lexer::SkipWhitespaces()
{
    while (!IsEof() && IsWhiteSpace(Peek()))
    {
        Forward();
    }
}